src/QsCompiler/Transformations/TypeResolutionSummary.cs

```diff
@@ -0,0 +1,390 @@
+
+        /// <summary>
+        /// Flag for if, at any time in the creation of the summary, there was a type parameter that
+        /// was assigned a type resolution referencing a different type parameter of the same callable.
+        /// </summary>
+        public bool SummarizesOverParameterConstriction { get; private set; } = false;
+
+        // Constructors
+
+        /// <summary>
+        /// Creates a type parameter resolution summary from the independent type parameter resolutions
+        /// found in the given typed expression and its sub expressions. Only sub-expressions whose
+        /// type parameter resolutions are relevant to the given expression's type parameter resolutions
+        /// are considered.
+        /// </summary>
+        public TypeResolutionSummary(TypedExpression expression) : this(GetTypeParameterResolutions.Apply(expression))
+        {
+        }
+
+        /// <summary>
+        /// Creates a type parameter resolution summary from independent type parameter resolution dictionaries.
+        /// The given resolutions are expected to be ordered such that dictionaries containing type parameters resolutions that
+        /// reference type parameters in other dictionaries appear before those dictionaries containing the referenced type parameters.
+        /// I.e., dictionary A depends on dictionary B, so A should come before B. When using this method to resolve
+        /// the resolutions of a nested expression, this means that the innermost resolutions should come first, followed by
+        /// the next innermost, and so on until the outermost expression is given last.
+        /// </summary>
+        internal TypeResolutionSummary(params TypeParameterResolutions[] independentResolutionDictionaries)
+        {
+            // Filter out empty dictionaries
+            this.ResolutionsInSummary = independentResolutionDictionaries.Where(res => !res.IsEmpty).ToImmutableArray();
+
+            if (!this.ResolutionsInSummary.Any())
+            {
+                this.CombinedTypeParameterResolutions = TypeParameterResolutions.Empty;
+            }
+
+            this.CombineTypeResolutions();
+        }
+
+        // Methods
+
+        /// <summary>
+        /// Updates the SummarizesOverParameterConstriction flag. If the flag is already set to true,
+        /// nothing will be done. If not, the given type parameter will be checked against the given
+        /// resolution for type parameter constriction, which is when one type parameter is dependent
+        /// on another type parameter of the same callable.
+        /// </summary>
+        private void UpdateConstrictionFlag(TypeParameterName typeParamName, ResolvedType typeParamResolution)
+        {
+            this.SummarizesOverParameterConstriction = this.SummarizesOverParameterConstriction
+                || ConstrictionCheck.Apply(typeParamName, typeParamResolution);
```

Comment:
Is it ok to short circuit the constriction check?

Reply:
The ConstrictionCheck is only there to set the `SummarizesOverParameterConstriction` flag. If we already know that there is a constriction violation, there is no need to continue to look for further constriction violations.